### DISCUSSION TITLE
AGPL compliance issue

### DISCUSSION THREAD

**Discussion by marcan** [minio/minio#12156] [Q&A] [answered]
Hi MinIO folks,

I learned that MinIO switched to the AGPL, and I was curious to see how the project achieves AGPL compliance, which is notoriously difficult and underdocumented for projects which do not have a sole copyright owner (only common for commercial software with open source dumps and with some oddball free projects that require copyright assignment, like the FSF's own). Unfortunately, given the AGPL as written, it would seen MinIO is currently in violation of its own license. More specifically, all MinIO contributors after the license change are in violation of all prior contributors' copyright.

Commit 06943256 switched the license of the project to AGPLv3. While it is unclear exactly how this relicensing affects the existing Apache licensed codebase, it is quite unambiguous that any further contributions after this point would be licensed under the AGPLv3.

The next commit, c829e3a1, was co-authored by @krisis, @poornas, and @krishnasrinivas. The code introduced by this commit would be licensed under the AGPLv3, and unambiguously make the derivative work it creates AGPLv3-only. As MinIO does not require copyright assignment for contributions, the overall codebase at this point is a derivative work authored by many people, and these three people have a copyright interest in the work as licensed under the AGPLv3.

In the next commit, 4d53054f, @harshavardhana modified an AGPLv3 program. As they do not own the copyright to the entire program, they are bound to follow its license when making modifications to the copyrighted work it covers.

Clause 13 of the AGPLv3 reads as follows:

> 13. Remote Network Interaction; Use with the GNU General Public License.
> 
> Notwithstanding any other provision of this License, if you modify the Program, your modified version must prominently offer all users interacting with it remotely through a computer network (if your version supports such interaction) an opportunity to receive the Corresponding Source of your version by providing access to the Corresponding Source from a network server at no charge, through some standard or customary means of facilitating copying of software. This Corresponding Source shall include the Corresponding Source for any work covered by version 3 of the GNU General Public License that is incorporated pursuant to the following paragraph.

Therefore, the modified version authored by @harshavardhana must "prominently offer all users interacting with it remotely [...] an opportunity to receive the Corresponding Source".

This offer is not in place. Building and running MinIO yields no source code offer when accessed via `curl`:

```
$ curl -i http://192.168.3.10:9000
HTTP/1.1 403 Forbidden
Accept-Ranges: bytes
Content-Length: 226
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Server: MinIO
Vary: Origin
X-Amz-Request-Id: 167944E3B8BA24E3
X-Xss-Protection: 1; mode=block
Date: Mon, 26 Apr 2021 01:41:25 GMT

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied.</Message><Resource>/</Resource><RequestId>167944E3B8BA24E3</RequestId><HostId>642b02d4-039e-427c-9719-06764cf33e91</HostId></Error>
```

Nor via the web browser:

![image](https://user-images.githubusercontent.com/98387/116018282-ae93a800-a67c-11eb-8f7f-81830d41a2d7.png)

In both cases references to MinIO being the server exist, but there is no link or reference to a network server from which the user may receive the source code, nor is there any reference to the exact source code version used to build MinIO. I do not see how this can meet the requirements of the license.

**Comment by harshavardhana** [comment, marked as answer]
Please email us at legal@minio.io or hello@minio.io for specific discussions such as these @marcan.  Closing this issue as the interpretation here is not legally vetted requires a more offline analysis.

> Therefore, the modified version authored by @harshavardhana must "prominently offer all users interacting with it remotely [...] an opportunity to receive the Corresponding Source".

The source code is directly available here at https://github.com/minio/minio

**Reply by Antidote** [reply]
Linking the repository doesn't resolve the issue @marcan is bringing up, AGPL requires an *in-app* way to request the source code running on the server, not the unmodified project repository.

**Reply by delroth** [reply]
This "interpretation" is also what the FSF describes [in their own FAQ](https://www.gnu.org/licenses/gpl-faq.html#AGPLProxy).

> **For software that runs a proxy server licensed under the AGPL, how can I provide an offer of source to users interacting with that code?**
> 
> For software on a proxy server, you can provide an offer of source through a normal method of delivering messages to users of that kind of proxy. For example, a Web proxy could use a landing page. When users initially start using the proxy, you can direct them to a page with the offer of source along with any other information you choose to provide.

This FAQ directly clarifies the FSF's stance with an example (a modified AGPL proxy server, which would theoretically have to insert an interstitial landing page with an offer of source code when a user is detected to first have hit the proxy). This clearly indicates that just having a tarball or a git repository online isn't the intention of the license.

**Reply by d235j** [reply]
There are two issues here. The first is the lack of an offer of source. The second is the lack of a CLA for external contributions. I would argue that the second issue is significantly more problematic, but I'll address these in order.

For the issue of the offer of source:
My suggestion is to have a parameter in the configuration that controls a string that appears in both the HTTP responses and the webpage providing the offer of source. This string can by default be a link to the GitHub or the website, but needs to be user configurable in case the user or the Linux distro has applied additional patches. For instance, it could say "mini version/build [vers]. Please submit AGPLv3 source code requests to [email-address]", or "minio version/build [vers], AGPLv3 source code is available at [URL]." The documentation also needs to make it clear that a user or distro that applies their own patches is responsible for updating this offer of source to allow requesting or obtaining their patched version.

The lack of CLA is a bigger problem. 
Without a CLA, if you accept pull requests or contributions from third parties, those contributions are copyright by the third party, and licensed only under the AGPL. Therefore you cannot offer minio containing those contributions under a commercial license at all! Most large open source projects that have commercial versions, such as Qt, Ubuntu, and others, require contributors to agree to a Contributor License Agreement (CLA) which either transfers the copyright to them, or provides a license to them that allows for redistribution under a license other than the open source license.

That said: none of this is legal advice. Please consult with a lawyer who has expertise in open source licensing for further assistance.

**Reply by marcan** [reply]
The config option thing doesn't work, because configuration is not part of the app itself (it is not copyrightable), and the AGPL imposes no requirements on users who do not modify the (copyrightable) application. If that approach worked, it would be a universal AGPL bypass:

1. I patch an app to add or transform the source code offer into using a configuration option, as well as any other changes I want to make, and give it to you
2. You host the app, with the configuration set to the wrong or no source code URL.

In this scenario, if the configuration option solution satisfied AGPL requirements, nobody has violated the license. So we have to conclude that for the AGPL to be useful, the source code offer URL has to be hard-coded, and a mere configuration option cannot satisfy the clause.

**Comment by SoniEx2** [comment]
1. Wouldn't the upstream be the thing that matters? If it's in the upstream, that makes it an official version, so clause 13 wouldn't apply.
2. Doesn't intent (i.e. intent of contributing to upstream vs intent of having ppl run the code) matter? After all, something that never runs can't be interacted with, and if your fork is not meant for running...

IANAL but we think these make it fine.

**Reply by marcan** [reply]
"Upstream" is not a legal concept. The AGPL makes no distinction of that nature. For a project without copyright assignment requirements, every contributor is a copyright owner, and thus every subsequent contributor has to follow all the licensing requirements and clauses in order to not violate every prior contributor's copyright.

The "upstream doesn't have to care about this" thing only works if your project has copyright assignment. Like, funny enough, many of the FSF's projects do.

"Intent" doesn't really matter; if you are distributing the code, someone can obviously run it. The license triggers on modification, it doesn't care what your plan is with those modifications. The entire way the AGPL works is based on this being enforced rather strictly so modified versions without a good source code offer cannot possibly exist, because once they do, thanks to Clause 9, anyone can use them without any obligations, including host them as a SaaS.

**Reply by SoniEx2** [reply]
At least for our own projects, based on our own interpretation, the AGPLv3 section 13's network interaction part does not apply for various reasons, and in fact the use of AGPLv3 mixed with GPLv3 creates enough additional loopholes that you don't even need to release most source when using our projects. Our projects also don't have an upstream, tho, so they're probably a better fit for the AGPLv3 in the first place...

But we digress. We see no reason to pretend the upstream doesn't exist for projects that do have a clearly-defined upstream. We also believe intent does matter because a version that doesn't run (or, alternatively, isn't running) cannot be interacted with, and only changes to get it to run would fall under section 13, thus being the responsibility of the person trying to run the fork instead.

**Reply by Antidote** [reply]
Your interpretation is flawed and goes against a literal reading of the text in both Clause 9 and 13, I'm sure your lawyers would agree with marcan.

**Reply by SoniEx2** [reply]
Do you mean the part about our projects or the part about the AGPLv3 in general? Our projects aren't even intended to run on a server, and they don't, themselves, interact with the network.

**Reply by Antidote** [reply]
It's still interacted with, which means you need to provide a clear means to acquire the source code, regardless of how it's presented, as dictated by the AGPL.

**Reply by SoniEx2** [reply]
Yeahhh no you don't. If the Program isn't even running, how can you be interacting with it?

(Also maybe we should move this to https://github.com/ganarchy/GAnarchy at this point.)

**Reply by Antidote** [reply]
Then why does this project even exist? Just toss it if it's not meant to be run. Your argument makes absolutely no sense.

**Reply by SoniEx2** [reply]
Oh it runs and then it stops running, it's just not something you can interact with while it's running. It's non-interactive.

**Reply by marcan** [reply]
For non-interactive programs without network features, yes, there is no requirement for a source code offer in the AGPLv3.

That aside, no, there is no "pretending an upstream doesn't exist". Upstream is not a legal concept. The legal concept is copyright, and copyright is owned by individual contributors in non-CLA project. There is no "upstream" legal entity and whoever controls the "upstream" repo has no special legal rights. Granting "upstream" special permission is *precisely* what a CLA does and why it is needed. No CLA, no "upstream" to speak of.

If a derivative work of an AGPLed project does not, indeed, function, then indeed there are no special requirements on it. For a contrived example, it is ostensibly legal to fork an AGPL project, add a typo (so it doesn't compile), then make any other changes, and not update any source code offer features. But then, if anyone fixes the typo to get it to run, they are responsible for simultaneously updating the source code offer feature. If a version of the project *can* run then the person who authored that derivative work *is* responsible for this. It doesn't matter if anyone actually runs it or not. It is the existence of the derivative work containing network interaction features that triggers the clause, not anyone else actually using it.

**Reply by SoniEx2** [reply]
If someone makes a compiler that automatically fixes typos, does that suddenly make it illegal? This is why intent (should) matter. But also, trademarks are a thing, so how do you pretend upstream doesn't exist when those are involved?

(Then again if GAnarchy actually succeeds copyright as we know it will become irrelevant.)

**Reply by marcan** [reply]
Compilation creates a derivative work, so the person running the compiler would be fixing the program and thus be on the hook at that point.

Trademarks have nothing to do with any of this. That is a completely different section of law, and does not interact with copyright in any way for our purposes. Copyright law doesn't care whether we have some idea of what "upstream" is through any other means. It only cares about copyright ownership and rights, and in those terms, "upstream" is no different from any other contributor.

**Reply by SoniEx2** [reply]
Granted, the best way to kill copyright is to actually kill copyright than to make bandaid licenses.

**Reply by SoniEx2** [reply]
Hmm... In some sense, the project name and details can be argued to be a strong enough pointer to the upstream for the purposes of identifying where to find the sources... So the upstream does play a part in it. It's not like the license explicitly calls out for the use of an URL, after all.

**Reply by marcan** [reply]
The license says "prominently offer all users interacting with it remotely through a computer network (if your version supports such interaction) an opportunity to receive the Corresponding Source of your version by providing access to the Corresponding Source from a network server at no charge".

I don't think you want to bet your license compliance on being able to successfully argue before a judge that having the project name is "prominent", and that the mere mention qualifies as an "offer" at all, or that users having to search for the correct Corresponding Source themselves is sufficient.

**Reply by SoniEx2** [reply]
You (the project developers) can write a clarification saying it, just like the linux kernel has a clarification about syscalls and whatnot. Whether that'd please everyone, probably not, but, it's an option?
